### ISSUE TITLE
better error messages with context

### ISSUE THREAD

**Issue by Marwe** [jgm/pandoc-citeproc#406]
Using pandoc with pandoc-citeproc often yields errors that are hard to trace back to source (=bibtex) files. The error messages in pandoc look like this:

> pandoc-citeproc: "stdin" (line 241, column 2):
> unexpected "M"
> expecting "c", "C", "p", "P", "s" or "S"
> CallStack (from HasCallStack):
>   error, called at src/Text/CSL/Input/Bibtex.hs:121:32 in pandoc-citeproc-0.10.5.1-HIfBMT1iu3x9mOf0xuZHLo:Text.CSL.Input.Bibtex
> pandoc: Error running filter pandoc-citeproc
> Filter returned error status 1

Since I have multiple, large bibtex files, that I generate automatically and include with --bibliography= to pandoc, it is hard to find the error, since the context of the parsing fail is missing: 
* stdin > so there is no filename
* no context of the failing bibtex entry text

Finding and correcting failing/erroneus entries is very hard this way.
I would like to see some of the parse context (lines before/after failing read) to identify the bibtex entry based on its content (e.g. based on bibkey, title, authors or any other fields).

Basically any error in a bibfile can lead to a bail out. I might currently be seeing errors due to @comments (as in #306), but I am not sure about it.

Cleaning up bibfiles is more or less luck, and usually only feasible, if the latest added entries fail and therefore can be found by log/date/diff.

I usually call pandoc from script or commandline with quite a few options, at least like this and with multiple bibliography options:
`pandoc --latex-engine=xelatex -fmarkdown+autolink_bare_uris+lists_without_preceding_blankline --template=/path/to/t.tex --bibliography=/path/to/bib --csl /path/to/csl -o out.html in.md`

**Comment by jgm** [comment]
What version of pandoc?
With recent versions you should be getting the filename, at least.

**Comment by Marwe** [comment]
OK, my pandoc was a bit outdated (installed via package manager on ubuntu 18.04: [bionic:pandoc packages](https://packages.ubuntu.com/search?keywords=pandoc&searchon=names&suite=bionic&section=all)): pandoc 1.19.2.4, pandoc-citeproc 0.10.5.1

With a freshly compiled pandoc-citeproc (pandoc 2.2.1, pandoc-citeproc 0.16.0.2) the filename of the bibfile is given and that points to the error precisely, just as you explained.
So an upgrade to a recent version solves this issue, sorry for the noise.


I just stumbled across a different error, where a filename is not given. Bailout due to non-utf8 encoded bibfiles just names the reason, not the file: 

> pandoc-citeproc: Cannot decode byte '\xf6': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream
>  # return/exit value 83

OK, encoding could be checked with external tools (e.g. `file` guesses it), if you have the list of input files (in my case a bash script adds multiple bibfiles, so I can look at the verbose output of the bash execution). Anyway, it would be nice to have a kind of verbose flag for pandoc/citeproc that lists the processed file(s) or input. This might be a different issue, since the original issue is solved.

Thank you for looking into this and the great software!

**Comment by jgm** [comment]
It's a different issue. If you want to create a separate issue for this, go ahead.
